infrastructure/templates/public-api/components/postgresqlDatabase.bicep

```diff
@@ -0,0 +1,153 @@
+
+@description('Geo-Redundant Backup setting')
+param geoRedundantBackup string = 'Disabled'
+
+@description('The name of the Key Vault to store the connection strings')
+param KeyVaultName string
+
+//Passed in Tags
+param departmentName string = 'Public API'
+param environmentName string = 'Development'
+param solutionName string = 'API'
+param subscriptionName string = 'Unknown'
+param costCentre string = 'Unknown'
+param serviceOwnerName string = 'Unknown'
+param dateProvisioned string = utcNow('u')
+param createdBy string = 'Unknown'
+param deploymentRepo string = 'N/A'
+param deploymentScript string = 'N/A'
+
+// Variables and created data
+var metadataDatabaseName = '${subscription}-postgre-${serverName}'
+var pythonDbConnectionString = 'database=<database>, user=${administratorLoginName}@${metadataDatabase.name}, host=${metadataDatabase.name}${environment().suffixes.sqlServerHostname}, password=${administratorLoginPassword}, port=5432'
+var adoNetDbConnectionString = 'Server=${metadataDatabase.name}${environment().suffixes.sqlServerHostname};Database=<database>;Port=5432;User Id=${administratorLoginName}@${metadataDatabase.name};Password=${administratorLoginPassword};'
+
+var pythonSecretName = 'pythonConnectionString'
+var adoNetSecretName = 'adoConnectionString'
+
+
+//Resources 
+resource metadataDatabase 'Microsoft.DBforPostgreSQL/flexibleServers@2023-03-01-preview' = {
+  name: metadataDatabaseName
+  location: location
+  sku: {
+    name: skuName
+    tier: skuTier
+  }
+  properties: {
+    createMode: 'Default'
+    version: postgresqlVersion
+    administratorLogin: administratorLoginName
+    administratorLoginPassword: administratorLoginPassword
+    storage: {
+      storageSizeGB: storageSizeGB
+      autoGrow: autoGrowStatus
+    }
+    backup: {
+      backupRetentionDays: backupRetentionDays
+      geoRedundantBackup: geoRedundantBackup
+    }
+    highAvailability: {
+      mode: 'Disabled'
+    }
+  }
+  tags: {
+    Department: departmentName
+    Solution: solutionName
+    ServiceType: 'PostgreSQL Database'
+    Environment: environmentName
+    Subscription: subscriptionName
+    CostCentre: costCentre
+    ServiceOwner: serviceOwnerName
+    DateProvisioned: dateProvisioned
+    CreatedBy: createdBy
+    DeploymentRepo: deploymentRepo
+    DeploymentScript: deploymentScript
+  }
+}
+
+
+//store connections strings
+module StorePythonConnectionStringToKeyVault './keyVaultSecret.bicep' = {
+  name: 'python-kv-connectionString'
+  params: {
+    KeyVaultName: KeyVaultName
+    IsEnabled: true
+    SecretValue: pythonDbConnectionString 
+    ContentType: 'text/plain'
+    SecretName: pythonSecretName
+  }
+}
+
+module StoreADOConnectionStringToKeyVault './keyVaultSecret.bicep' = {
+  name: 'ado-kv-connectionString'
+  params: {
+    KeyVaultName: KeyVaultName
+    IsEnabled: true
+    SecretValue: adoNetDbConnectionString 
+    ContentType: 'text/plain'
+    SecretName: adoNetSecretName
+  }
+}
```

Comment:
As discussed, we probably shouldn't override the secret if one hasn't already been deployed to Key Vault.

To do the initial secret creation, we would need to pass in an generated random string for the admin password as a pipeline parameter. There'd be two ways of doing this:

1. Generating it a new one on each deployment
2. Just set a generated password in manually

With option 1, we'd need to be careful that this doesn't try and change the database's admin password every time there's a deployment.

With option 2, we'd need to remove the password once it's been made available to Key Vault.

Thinking about it some more though, I think we might want to consider some alternatives...

## Deploying Key Vaults with 'seed' secrets

This would involve creating a separate Bicep deployment that only deploys the Key Vaults and any 'seed' secrets, such as the admin password.

These Key Vaults could then be referenced in this deployment to extract the admin password and use it for the database and generating the connection string secret.

## Using Managed Identities

I think this is arguably the best option. Passwordless authentication using Managed Identities is pretty ideal and would mean we don't need to go around the houses with this.

I'm sure we'll probably run into permission issues in the DfE Azure, but I think this is probably the way to go.

Given that we already have to make a managed identity for any Container App, this doesn't feel like too much more on top!

With this approach, we'll also need to configure the Processor (ETL) with an appropriate managed identity.